dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/TaskSetState.java

```diff
@@ -16,60 +16,134 @@
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
+    @Override
+    public synchronized long failedRequiredTaskCount() {
+      long result = 0;
+      for (Map.Entry<Task<?>, TaskResult<?>> entry : resultMap.entrySet()) {
+        TaskState state = entry.getValue().getState();
+        if (entry.getKey().getCategory() == REQUIRED && state == FAILED) {
+          result++;
+        }
+      }
+      return result;
+    }
 
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
     @Nonnull
-    public Map<Task<?>, TaskResult<?>> getTaskResultMap() {
-      synchronized (lock) {
-        return resultMap;
-      }
+    @Override
+    public synchronized ImmutableList<TaskResultSummary> taskResultSummaries() {
+      return resultMap.entrySet().stream()
+          .map((entry) -> TaskResultSummary.create(entry.getKey(), entry.getValue()))
+          .collect(toImmutableList());
     }
 
+    @Deprecated // Use TaskSetState instead of TaskSetState.Impl
+    @Nonnull
     @Override
-    @SuppressWarnings("unchecked")
-    public <T> TaskResult<T> getTaskResult(@Nonnull Task<T> task) {
-      synchronized (lock) {
-        return (TaskResult<T>) resultMap.get(task);
-      }
+    public synchronized ImmutableList<TasksReport> tasksReports() {
+      return resultMap.values().stream()
+          .collect(groupingBy(TaskResult::getState, counting()))
+          .entrySet()
+          .stream()
+          .map((entry) -> new TasksReport(entry.getValue(), entry.getKey()))
+          .collect(toImmutableList());
     }
 
-    public <T> void setTaskResult(
+    @Nonnull
+    @Override
+    public TaskState getTaskState(@Nonnull Task<?> task) {
+      TaskResult<?> result = resultMap.get(task);
+      return (result == null) ? TaskState.NOT_STARTED : result.getState();
+    }
+
+    public synchronized <T> void setTaskResult(
         @Nonnull Task<T> task, @Nonnull TaskState state, @CheckForNull T value) {
-      synchronized (lock) {
-        resultMap.put(task, new TaskResult<>(state, value));
-      }
+      resultMap.put(task, new TaskResult<>(state, value));
     }
 
-    public <T> void setTaskException(
-        @Nonnull Task<T> task, @Nonnull TaskState state, @CheckForNull Exception exception) {
-      synchronized (lock) {
-        resultMap.put(task, new TaskResult<>(state, exception));
-      }
+    public synchronized void setTaskException(
+        @Nonnull Task<?> task, @Nonnull TaskState state, @CheckForNull Exception exception) {
+      resultMap.put(task, new TaskResult<>(state, exception));
     }
   }
 
-  @CheckForNull
-  public <T> TaskResult<T> getTaskResult(@Nonnull Task<T> task);
+  long failedRequiredTaskCount();
 
   @Nonnull
-  public default TaskState getTaskState(@Nonnull Task<?> task) {
-    TaskResult<?> result = getTaskResult(task);
-    return (result == null) ? TaskState.NOT_STARTED : result.getState();
-  }
+  ImmutableList<TaskResultSummary> taskResultSummaries();
```

Comment:
Should these be 'get' methods?

Reply:
Done, I added a `get-` prefix to the new methods.